Cmarley condensed robot code

cmarley3-14 opened the pull request:
### Changes Made:
**Lifter:**  
Changed the comparison tests so that it would see if the distance was above the target height or below, and move accordingly, instead of testing if the lifter was at a certain place. Because once the lifter is in transition, it wouldn't have fit any of the cases.

**LifterHight:**
Changed the height variables to `public` because they now need to be accessed in `Lifter.java` And then added extra spacing for the brackets and if statements (re-formatting).

**Robot:**
Created `AutonomousPeriodic`, which is part of the FIRST library. Added the drive controls, suck grab eject, and left climb out.  
Moved the code for lifting (The POV and switch statement stuff) out of `TeleopPeriodic()` and into its own function `lifterOperatorCode()`, because there's two exact same copies in both auto and teleop. Just to make the code shorter.
Changed `Climb()` to a series of steps which progress by pressing start. This is because with `Climb()`, one assumes that the robot will move backwards onto the platform in the short amount of time given by the quick succession of steps in such a function. The steps are more like:
```
raise robot
walk (the robot legs are angled such that the robot falls onto the platform)
retract back legs
drive backward a sufficient distance so the robot won't fall off the platform when...
the front legs are retracted
```
The driving portion of the steps was not accounted for, so the code was changed to compensate for that.

And that's the summary of the changes!

cmarley3-14 committed:
Moved Lifter Interface code to new function

cmarley3-14 committed:
See lines 62 - 65

The tests for lifter movement may fail.

cmarley3-14 committed:
Edited Climb() function

see comments

cmarley3-14 committed:
added m_lifter.stop

because otherwise we'll have no way to do motor.set(0); for the lifter winch. It'll be stuck always on +/-

cmarley3-14 committed:
added Stop()

because the lifter motor doesn't auto reset.

Things aren't idealized in the robot world.

cmarley3-14 committed:
code formatting and variables

have to change the hight variables to public because they'll need to be accessed in lifter
and re-tabbed and re-bracketed the code

cmarley3-14 committed:
changed conditions for lifter movement

direct testing of get distance for go to level functions

cmarley3-14 committed:
deprecation of Climb()

changes made to compensate in Robot.java()

cmarley3-14 committed:
deprecation of Climb()

changes made to compensate in Robot.java()

sprice134 commented:
It looks good to me, Id say we just gotta test it on the robot to make sure everything is in order. Only concern is more of a general concern than a specific coding change but making sure that the climbing function is fully completed before moving on, and possibly preventing a driver from moving to phase 2 (or whatever the next phase is) prior to completing the previous one.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Lifter.java:
```diff
@@ -42,75 +42,92 @@ public void Lower() {
+        }
     }
 
+    // All these go to level functions would be much easier in python.
```
Useless commentary. Delete.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Lifter.java:
```diff
@@ -42,75 +42,92 @@ public void Lower() {
-        }else if (m_hight.atTop()){
-            m_lifter.set(-kMotorPowerLevel); }  
+        } else if (m_hight.getDistance() < m_hight.ballThirdLevelHight) {
+            m_lifter.set(kMotorPowerLevel);
```
Note that "height" is spelled incorrectly here and basically everywhere else in the code.
But it can be fixed as a separate changelist. The filename is spelled wrong as well.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/LifterHight.java:
```diff
@@ -20,13 +20,13 @@
-    private static final double ballFirstLevelHight = 67.945;
-    private static final double ballSecondLevelHight = 100;
-    private static final double ballThirdLevelHight = 150;
+    public static final double ballFirstLevelHight = 67.945;
```
These should remain private. No other class needs to know the exact values here.

garykac replied:
Also, what are the units here? Centimeters? There should be a comment noting that info.

cmarley3-14 replied:
They need to be public.

The `Lifter` class will use these values in a bunch of comparison tests with `getDistance()` to determine how to move to get to various game piece levels.

`LifterHight` is basically a constants and boolean functions class for the actual lifter class. Nothing is wrong with keeping them public.

As for units... check with Ryan on that. This code is untested.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Robot.java:
```diff
@@ -31,22 +32,26 @@
-    private int lifterLevelHatch = 0;
-    private int lifterLevelBall = 0;
+    private int lifterLevel = 0;
+    private int modeAdder = 1; // You'll see its ingenuity later on... *wink*
```
This comment adds no value. Either change the comment to actually describe what the variable is doing, or delete it.

garykac replied:
Class member variable names should begin with `m_`

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Robot.java:
```diff
@@ -91,19 +99,80 @@ public void robotInit() {
     }
-
+    
+/** For line # 119
```
Referring to a specific line number is not useful when lines can be added/deleted.
Also this entire comment block is indented oddly.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Robot.java:
```diff
@@ -115,103 +184,48 @@ public void teleopPeriodic() {
-            ballMode=true;
-        }
+        } 
+        lifterOperatorCode(); // Will this work?
```
Comment adds no value. If you're not sure if the code will work, then there should be an issue tracking it, because no one is going to remember to come back and look at this specific line in the code.

cmarley3-14 replied:
There is an issue open for it already. The comment is for debugging, if we need to revert back to previous methods.

cmarley3-14 replied:
This has been marked as issue #25 for the past 3 or so days.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Walker.java:
```diff
@@ -41,6 +41,8 @@ public void RetractBackLegs() {
     }
 
+    // This function has been deprecated because it's too idealized
+    // The changes have been adapted in Robot.java
```
I don't understand what is meant by these comments. "too idealized"? If the code doesn't work than it should be fixed here. Moving the code into Robot does not seem necessary.

cmarley3-14 replied:
idealized meaning unrealistically perfect. See comments on proposed changes to lines 215 to 222.

For reference *(taken from said comment reply)*:
> The steps for climbing are:
1- drive backwards to platform
2- raise robot
3- "walk", where the back legs are tilted so that the robot falls backwards onto the platform the robot should fall back enough such that the wheels hit the platform
4- retract back legs, in preparation for driving backwards
5- drive backwards until robot is fully above platform
6- retract front legs. front wheels of robot will fall and rest on the platform edge

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Robot.java:
```diff
@@ -115,103 +184,48 @@ public void teleopPeriodic() {
+            if (macroIndex == 2)
+                m_walker.Walk();
+            if (macroIndex == 3) //By idealized, I mean, the robot has to drive during this step.
+                m_walker.RetractBackLegs(); // Climb() doesn't give the operator time to do so.
```
Does it need to? Retracting the legs should just initiate the process and then return control. It doesn't wait for the retraction to complete.  I'm not sure what problem this is trying to solve.

garykac replied:
Also, what happened to the RaiseFrontLegs step? I think this code is less likely to work than the original.

cmarley3-14 replied:
The steps for climbing are:
1- drive backwards to platform
2- raise robot
3- "walk", where the back legs are tilted so that the robot falls backwards onto the platform the robot should fall back enough such that the wheels hit the platform
4- retract back legs, in preparation for driving backwards
5- drive backwards until robot is fully above platform
6- retract front legs. front wheels of robot will fall and rest on the platform edge

the Climb() function basically went from one step to the next without waiting in between. It wouldn't have even finished waiting for the function of extension/retraction to execute. If you look at the 2018 code, you'll see a macro timer, which was necessary because each step took time to complete. Separating the Climb() function into its steps allows the operator to climb in his own timing.

cmarley3-14 replied:
This is discussed in issue #24.
https://github.com/garykac/amhs-robotics-4681/blob/2aa4fd3c6ba7218b489a7fc224bba6bec184faae/workspace/2018-FIRST_Power_Up/2018RampBot/src/org/usfirst/frc/team4681/robot/Robot.java#L331-L335
https://github.com/garykac/amhs-robotics-4681/blob/2aa4fd3c6ba7218b489a7fc224bba6bec184faae/workspace/2018-FIRST_Power_Up/2018RampBot/src/org/usfirst/frc/team4681/robot/Robot.java#L348-L382

This is the relevant 2018 robot code that demonstrates why we can't have a single climb function. Each step needs to be carried out separately.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Robot.java:
```diff
@@ -91,19 +99,80 @@ public void robotInit() {
+                    break;
+            }
+        } else { // The only other choice is autoLift = false;
+            if(m_stick.getPOV() == 0){
```
space after `if`

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Lifter.java:
```diff
@@ -42,75 +42,92 @@ public void Lower() {
-            m_lifter.set(kMotorPowerLevel); 
-        }else if (m_hight.atSecondBallLevel() || m_hight.atThirdBallLevel()){
-            m_lifter.set(-kMotorPowerLevel); }  
+        } else if (m_hight.getDistance() < m_hight.ballFirstLevelHight) {
```
This is MUCH better than the previous code.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/LifterHight.java:
```diff
@@ -36,57 +36,81 @@ public void lifterHightInit() {
         
     }
+    
     public double getDistance(){
```
What units does this return?

cmarley3-14 replied:
I don't think the code has been tested. We've had quite a few snow days, sadly...

cmarley3-14 replied:
See issue #27 for a reference to these and other problems raised.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/LifterHight.java:
```diff
@@ -36,57 +36,81 @@ public void lifterHightInit() {
         if (counter.get() < 1) {
             return 0.0;
         }
         return (counter.getPeriod() * 1000000.0 / 10.0) + offSet;
```
"1000000.0 / 10.0" is the same as "100000.0".  Where did these numbers come from?

cmarley3-14 replied:
This is Ryan's code. I don't think it has been tested, and I don't much see how it works.

cmarley3-14 replied:
See issue #27 for a reference to these and other problems raised.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/LifterHight.java:
```diff
@@ -36,57 +36,81 @@ public void lifterHightInit() {
-    public double minOffSet(double hight){
-        return (hight-minMaxDef);
+    public double minOffSet(double hight) {
+        return (hight - minMaxDef);
```
`minMaxDef` is "5". If this is cm, then our `atXxx()` methods below are accurate to +/- 5cm, which is a 10cm range - which seems large.

cmarley3-14 replied:
See issue #27 for a reference to these and other problems raised.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/LifterHight.java:
```diff
@@ -36,57 +36,81 @@ public void lifterHightInit() {
+        return false;
+    }
+    
+    public boolean atSecondHatchLevel() {
```
The code for all these `atXxx()` methods is basically the same, except for the value that they're checking.
This can be pulled out into a separate method and then all the `atXxx()` methods become a single line (calling this new method).

If that had been done earlier, then you would have only needed to fix the code in one place.

cmarley3-14 replied:
It can, and it will. I have time tomorrow (Tuesday)

cmarley3-14 replied:
It may as well be too late. The effects are too widespread to work with at this point.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Robot.java:
```diff
@@ -91,19 +99,80 @@ public void robotInit() {
+    
+/** For line # 119
+  * lifterLevel will either be offset by +1 (ball mode) or 0 (hatch mode). It starts at +1.
+  * The lifter levels are as follows: -1=bottom, 0=bottom, 1=hatch1, 2=ball1, 3=hatch2, 4=ball2, 5=hatch3, 6=ball3
```
How is this going to handle the levels for getting a hatch or a ball?

cmarley3-14 replied:
p. 18 of the Game Manual states:
> The HATCH PANEL placement height at the bottom of a ROCKET equals the HATCH
PANEL placement height on the CARGO SHIP (per the CARGO SHIP section), which
equals the HATCH PANEL retrieval height from the LOADING STATION (per the
LOADING STATION section). 

The lower limit of the cargo bay for the cargo ship is 80 cm above the ground. We can just add a new variable for that.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Lifter.java:
```diff
@@ -42,75 +42,92 @@ public void Lower() {
-            m_lifter.set(kMotorPowerLevel); 
-        }else if (m_hight.atThirdHatchLevel() || m_hight.atSecondHatchLevel()){
-            m_lifter.set(-kMotorPowerLevel); }  
+            loc = Location.ballHatchLevel;
```
`ballHatchLevel` -> `hatchFirstLevel`.  This code won't compile. You should always make sure it compiles before trying to check it in.

cmarley3-14 replied:
Okay. These two files will be revised tomorrow (Tuesday), so that the code is easier to understand.

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Lifter.java:
```diff
@@ -42,75 +42,92 @@ public void Lower() {
-            m_lifter.set(kMotorPowerLevel); 
-        }else if (m_hight.atThirdHatchLevel() || m_hight.atTop()){
-            m_lifter.set(-kMotorPowerLevel); }  
+            loc = Location.ballHatchLevel;
```
`ballHatchLevel` -> `hatchSecondLevel`

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Lifter.java:
```diff
@@ -42,75 +42,92 @@ public void Lower() {
-            m_lifter.set(kMotorPowerLevel); 
-        }else if (m_hight.atTop()){
-            m_lifter.set(-kMotorPowerLevel); }  
+            loc = Location.ballHatchLevel;
```
`ballHatchLevel` -> `hatchThirdLevel`

garykac commented on workspace/2019 DeepSpace/src/main/java/frc/robot/Lifter.java:
```diff
@@ -42,75 +42,92 @@ public void Lower() {
-            m_lifter.set(kMotorPowerLevel); 
-        }else if (m_hight.atThirdBallLevel() || m_hight.atTop()){
-            m_lifter.set(-kMotorPowerLevel); }  
+        } else if (m_hight.getDistance() < m_hight.ballSecondLevelHight) {
```
Rather than exposing these private values, this could instead be:
```
if (m_hight.atSecondBallLevel()) {
    m_lifter.set(0);
    loc = Location.ballSecondLevel;
} else {
    m_lifter.goToSecondBallLevel();
}
```

cmarley3-14 replied:
That becomes a recursive function. `m_lifter.GoToSecondBallLevel()` is what is being defined. It requires the lifter height locations to know whether to move up or down.

garykac reviewed (changes requested):
Thanks for cleaning up some of the code formatting problems.

cmarley3-14 committed:
took out the comment

cmarley3-14 committed:
Lines 67-68: Redefine functions

Given a bunch of variables e.g. `hatchLevelOneHight`, wouldn't it be easier to have a single function:
``` java
public void goTo(int specHeight) {
    if (m_hight.near(specHeight)) {
        m_lifter.set(0)
// etc.
```
where `m_hight.near( int specHeight2)` is just the minMaxOffset function?

cmarley3-14 committed:
updated enum names (necessary?)

if we aren't going to use the enums, why bother?

cmarley3-14 committed:
move out week 5, change week 6

cmarley3-14 committed:
added week 5

cmarley3-14 committed:
comments and functions revised

redid comment on `modeAdder`
added emergency restart for climber

cmarley3-14 committed:
more brackets and spaces changes

cmarley3-14 committed:
Merge branch 'gh-pages' into cmarley-condensedRobotCode
